src/binary-encoding.adoc

```diff
@@ -0,0 +1,1621 @@
+       bytes
+----
+
+[sidebar]
+TODO: Demonstrate splicing macro values into the struct via VarSym escape code `0x00`.
+
+[[delimited_structs]]
+==== Delimited Structs
+
+Opcode `0xF3` indicates the beginning of a delimited struct with <<varsym, `VarSym`>> field names,
+while opcode `0xF0` closes the most recently opened delimited container that has not yet been closed.
+
+NOTE: While length-prefixed structs can choose between <<structs_with_symbol_address_field_names, structs with
+symbol address field names>> and <<structs_with_varsym_field_names, structs with `VarSym` field names>>,
+delimited structs always use `VarSym`-encoded field names.
+
+.Figure {counter:figure}: Delimited encoding of the empty struct (`_{}_`)
+[source,%unbreakable]
+----
+┌─── Opcode 0xF3 indicates the beginning of a delimited struct with `VarSym` field names.│
+│  ┌─── VarSym escape code 0x00: an opcode follows
+│  │  ┌─── Opcode 0xF0 indicates the end of the most
+│  │  │    recently opened delimited container
+F3 00 F0
+----
+
+.Figure {counter:figure}: Delimited encoding of `_{"foo": 1, $11: 2}_`
+[source,%unbreakable]
+----
+┌─── Opcode 0xF3 indicates the beginning of a delimited struct with `VarSym` field names.
+│
+│  ┌─ VarSym -3      ┌─ VarSym: 11 ($11)
+│  │                 │        ┌─── VarSym escape code 0x00: an opcode follows
+│  │                 │        │  ┌─── Opcode 0xF0 indicates the end of the most
+│  │   f  o  o       │        │  │    recently opened delimited container
+F3 FD 66 6F 6F 51 01 17 91 02 00 F0
+      └──┬───┘ └─┬─┘    └─┬─┘
+      3 UTF-8    1        2
+       bytes
+----
+
+[[nulls]]
+=== Nulls
+
+The opcode `0xEA` indicates an untyped null (that is: `null`, or its alias `null.null`).
+
+The opcode `0xEB` indicates a typed null; a byte follows whose value indicates the type. The byte-to-type
+mapping is as follows:
+
+[cols="^1a,4a"]
+|===
+|Byte |Type
+
+|`0x00`
+|`null.bool`
+
+|`0x01`
+|`null.int`
+
+|`0x02`
+|`null.float`
+
+|`0x03`
+|`null.decimal`
+
+|`0x04`
+|`null.timestamp`
+
+|`0x05`
+|`null.string`
+
+|`0x06`
+|`null.symbol`
+
+|`0x07`
+|`null.clob`
+
+|`0x08`
+|`null.blob`
+
+|`0x09`
+|`null.list`
+
+|`0x0A`
+|`null.sexp`
+
+|`0x0B`
+|`null.struct`
+|===
+
+.Figure {counter:figure}: Encoding of `_null_`
+[source,%unbreakable]
+----
+┌──── The opcode `0xEA` represents a null (null.null)
+EA
+----
+
+.Figure {counter:figure}: Encoding of `_null.string_`
+[source,%unbreakable]
+----
+┌──── The opcode `0xEB` indicates a typed null; a byte indicating the type follows
+│  ┌──── Byte 0x05 indicates the type `string`
+EB 0x5
+----
+
+[[annotations]]
+=== Annotations
+
+Annotations can be encoded either <<annotations_with_symbol_addresses, as symbol addresses>>
+or <<annotations_with_varsym_text, as ``VarSym``s>>. In both encodings, the annotations sequence appears
+just before the value that it decorates.
+
+If an annotations sequence appears before one or more additional annotations sequences, the sequences
+are concatenated.
```

Comment:
I don't think I like this from an implementation perspective... In the reader, I'd love to be able to scan for values and record one start/end position for its annotation sequence that I can return to later if the user requests annotations. It's less simple (and quite possibly slower) if we're supporting N annotation sequences between values. Is there a killer use case I'm missing?

Reply:
Allowing 'chaining' means you can write annotation sequences of arbitrary length without buffering. However, in practice sequences are so rarely longer than 2 annotations (or ~4 bytes) that this probably isn't very useful. I'm ok with removing it.